apis/Google.Cloud.Metadata.V1/Google.Cloud.Metadata.V1/MetadataClientImpl.cs

```diff
@@ -0,0 +1,502 @@
+        {
+            add
+            {
+                HookChangeEvent(ref maintenanceStatusWatcher, value, "instance/maintenance-event", (sender, newValue) =>
+                {
+                    var temp = sender.Event as EventHandler<MaintenanceStatus>;
+                    temp?.Invoke(this, GetMaintenanceStatus(newValue));
+                });
+            }
+            remove
+            {
+                UnhookChangeEvent(ref maintenanceStatusWatcher, value);
+            }
+        }
+
+        /// <inheritdoc/>
+        public override event EventHandler ProjectMetadataChanged
+        {
+            add
+            {
+                HookChangeEvent(ref projectWatcher, value, "project?recursive=true", (sender, newValue) =>
+                {
+                    var temp = sender.Event as EventHandler;
+                    temp?.Invoke(this, EventArgs.Empty);
+                });
+            }
+            remove
+            {
+                UnhookChangeEvent(ref projectWatcher, value);
+            }
+        }
+
+        /// <inheritdoc/>
+        public override string WaitForChange(string key, TimeSpan timeout = default(TimeSpan)) => GetResult(token => WaitForChangeAsync(key, timeout, token));
+
+        /// <inheritdoc/>
+        public override async Task<string> WaitForChangeAsync(string key, TimeSpan timeout = default(TimeSpan), CancellationToken cancellationToken = default(CancellationToken))
+        {
+            // TODO: Validate the key
+
+            var uri = new Uri($"http://temp/{key}", UriKind.Absolute);
+            key += (string.IsNullOrEmpty(uri.Query) ? "?" : "&") + "wait_for_change=true";
+
+            if (!uri.Query.Contains("timeout_sec"))
+            {
+                if (timeout == TimeSpan.Zero)
+                {
+                    timeout = new TimeSpan(HttpClient.Timeout.Ticks / 2);
+                }
+
+                key += $"&timeout_sec={(int)timeout.TotalSeconds}";
+            }
+
+            var response = await RequestMetadata(key, cancellationToken).ConfigureAwait(false);
+            return await response.Content.ReadAsStringAsync().ConfigureAwait(false);
+        }
+
+        private class ValueWatcher
+        {
+            private readonly CancellationTokenSource cancellationTokenSource;
+            private readonly Action<ValueWatcher, string> changeAction;
+            private readonly MetadataClientImpl client;
+            private readonly string path;
+            private readonly SynchronizationContext syncContext;
+            private readonly Task waitTask;
+
+            public Delegate Event { get; set; }
+
+            public ValueWatcher(MetadataClientImpl client, string path, Action<ValueWatcher, string> changeAction)
+            {
+                this.changeAction = changeAction;
+                this.client = client;
+                this.path = path;
+
+                syncContext = SynchronizationContext.Current;
```

Comment:
This is really mostly for UI applications, for server applications the synchronization context doesn't do much. I am thinking that we can do away with it since none of the other libraries do this, and probably the only UI app that will ever do this is VS .